### ISSUE TITLE
Windows Support

### ISSUE THREAD

**Issue by smarts** [redox-os/ion#285]
**Reproduction**: `cargo build`

**Expected behavior**: a successful build

**Actual behavior**:
```
   Compiling permutate v0.2.0
   Compiling ansi_term v0.9.0
   Compiling termion v1.3.0
   Compiling peg v0.5.3
   Compiling smallstring v0.1.2
error[E0432]: unresolved import `libc::termios`
 --> {currentUserDirectory}\.cargo\registry\src\github.com-1ecc6299db9ec823\termion-1.3.0\src\termios.rs:4:9
  |
4 | pub use libc::termios as Termios;
  |         ^^^^^^^^^^^^^^^^^^^^^^^^ no `termios` in the root

error[E0432]: unresolved import `libc::ioctl`
  --> {currentUserDirectory}\.cargo\registry\src\github.com-1ecc6299db9ec823\termion-1.3.0\src\size.rs:42:9
   |
42 |     use libc::ioctl;
   |         ^^^^^^^^^^^ no `ioctl` in the root

error[E0432]: unresolved import `libc::STDOUT_FILENO`
  --> {currentUserDirectory}\.cargo\registry\src\github.com-1ecc6299db9ec823\termion-1.3.0\src\size.rs:43:9
   |
43 |     use libc::STDOUT_FILENO;
   |         ^^^^^^^^^^^^^^^^^^^ no `STDOUT_FILENO` in the root

error[E0432]: unresolved import `std::os::unix::io::AsRawFd`
 --> {currentUserDirectory}\.cargo\registry\src\github.com-1ecc6299db9ec823\termion-1.3.0\src\tty.rs:2:5
  |
2 | use std::os::unix::io::AsRawFd;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^ Could not find `unix` in `os`

error: cannot continue compilation due to previous error

error: Could not compile `termion`.
Build failed, waiting for other jobs to finish...
error: build failed
```

**Build information**: 
 - rustc 1.19.0-nightly (107bd67ef 2017-06-02)
 - i didn't clone the repo, just downloaded a zip file of `master` @ 29497ada7fcf656c30da89d9689769163913fdf1

**Misc**: OS is Windows 8 Pro

**Comment by mmstick** [comment]
There are three things that prevent Windows support:

- [ ] Termion does not support Windows
- [ ] Piping requires the unix module to gain access to raw file descriptors on unix systems
- [ ] As far as I know, none of us have or use Windows.

**Comment by smarts** [comment]
thanks for the reply! =) i opened this issue because i came across the issue i referenced above and it was closed because, as you mentioned, no one on the project uses Windows and therefore no one knew whether or not it worked on Windows. now it's clear that it doesn't.

**Comment by radix** [comment]
would it be useful to keep this ticket open with a list of blockers for Windows support? I would like a ticket to subscribe to so I can get notifications if the status changes.

**Comment by mmstick** [comment]
I'll create a new issue with the blockers, but I can tell you already that it's very unlikely that Windows will be supported any time soon. There's a lot of critical functionality needed by shells that Windows does not support at all, or the Windows way of implementing support is radically different from the *nix way. It would require significant time and effort in order to somehow replicate the missing functionality that Windows does not support, and re-implement all the shell functionality that Windows would need written from scratch using a different approach.

# Problems with Supporting Windows

Take the `fork` syscall, for example. The shell needs to be able to perform forking in order to create background processes that are an exact copy of the parent process but executing a different codepath, and to assign new process group IDs to these background child processes. Windows does not support forking as a concept at all, so we'd need to replicate the forking concept from scratch for Windows. It's one of the biggest reasons why most system shells are *nix-exclusive.

There's also the manner in which *nix architectures perform job control, of which Windows also does not support at all. With *nix architectures, you configure spawned processes to set their process group ID to be the same as their process ID, and manage processes with various *nix signals, and use functions like `tcsetpgrp` to set foreground processes.

Or the difference between how *nix systems operate on file descriptors, which are integers, but Windows operates on handles, which are void pointers. Jobs are controlled by passing along PIDs and mapping FDs on *nix systems, but Windows is very different.

# What Needs To Be Done

Basically, to gain Windows support, all of the code that manages foreground and background processes, process creation and execution, and piping / redirection would need to be rewritten from scratch with Windows-specific APIs. But for that to happen, we first need someone who uses and develops for Windows to be willing to do that.

**Comment by radix** [comment]
Thanks for the summary @mmstick.
